mpv: enable swift support

azuwis opened the pull request:
###### Description of changes

Since swift is available on master now, I've rebased @stephank's work on mpv https://github.com/stephank/nixpkgs/commit/4ce183cdd525da0f3cf4cec14356d056dbabed0e to master, and made this pull request for discussion.

When I used nix-tree to check the closure size of the built mpv, I've found swift lib references clang, and make the closure size pretty big:

<img width="1066" alt="Screenshot 2023-02-06 at 22 19 14" src="https://user-images.githubusercontent.com/9315/216995420-a7e094ef-ce12-4386-8cb9-a5058d328ed4.png">

```
$ ls -l /nix/store/2cp55whf6v3631512w4hlsrc4v7r8w36-swift-5.7.3-lib/lib/swift
total 0
dr-xr-xr-x  6 root nixbld 192 Jan  1  1970 FrameworkABIBaseline
dr-xr-xr-x  5 root nixbld 160 Jan  1  1970 _InternalSwiftScan
dr-xr-xr-x  5 root nixbld 160 Jan  1  1970 _InternalSwiftStaticMirror
dr-xr-xr-x  5 root nixbld 160 Jan  1  1970 _InternalSwiftSyntaxParser
dr-xr-xr-x  4 root nixbld 128 Jan  1  1970 apinotes
lrwxr-xr-x  1 root nixbld  78 Jan  1  1970 clang -> /nix/store/avrf44cq2cr18vlvv20aa1pc85cl7rzg-clang-wrapper-14.0.6/resource-root
dr-xr-xr-x 28 root nixbld 896 Jan  1  1970 macosx
dr-xr-xr-x 12 root nixbld 384 Jan  1  1970 migrator
dr-xr-xr-x 27 root nixbld 864 Jan  1  1970 shims
```

Is that expected?

Removing `preConfigure` will make the closure size even bigger:

<img width="1060" alt="Screenshot 2023-02-06 at 22 27 09" src="https://user-images.githubusercontent.com/9315/216997415-efafb94b-81ef-42f3-8ad9-5400177e7317.png">

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

stephank committed:
mpv: enable swift support

azuwis commented:
Related https://github.com/NixOS/nixpkgs/issues/160828 https://github.com/NixOS/nixpkgs/pull/204979

wegank commented on pkgs/os-specific/darwin/apple-sdk-11.0/frameworks.nix:
```diff
@@ -1,7 +1,7 @@
 {
   AGL                              = { inherit Carbon OpenGL; };
-  AVFoundation                     = { inherit ApplicationServices AVFCapture AVFCore CoreGraphics simd UniformTypeIdentifiers; };
+  AVFoundation                     = { inherit ApplicationServices AVFCapture AVFCore CoreGraphics CoreMedia simd UniformTypeIdentifiers; };
```
These changes look huge. Should the PR target staging?

wegank replied:
And I think #204979 does it better.

azuwis replied:
OK, took the approach of https://github.com/NixOS/nixpkgs/pull/204979, and force pushed.

azuwis committed:
mpv: Switch to apple_sdk_11_0 for darwin

wegank commented:
@ofborg build mpv mpv.passthru.tests

stephank commented:
Thanks for point out the closure size! It's something I'll make a PR for separately.

I'm seeing the same build failure as ofBorg on `x86_64-darwin`, but I don't understand it yet.

AndersonTorres commented:
Unfortunately (or not) I can't possibly comment, because this is too Darwin-specific to me.

Nonetheless, I wish good luck.

azuwis commented:
The error log on x86_64-darwin is different before and after commit https://github.com/NixOS/nixpkgs/pull/214944/commits/64bce1df1712a8731a810e1adf7281bce22038e8

Before(https://logs.nix.ci/?key=nixos/nixpkgs.214944&attempt_id=5ab1c3e7-d414-4bbe-9a9a-0bce635c1dc4):
```
[21/452] Generating generated/osdep/swift_targets with a custom command
FAILED: generated/osdep/macOS_swift.swiftmodule generated/osdep/macOS_swift.h generated/osdep/macOS_swift.o 
/nix/store/5mldrl986vi6b8lk3gdw8r5mjxqx5i9j-swift-wrapper-5.7.3/bin/swift -frontend -c -sdk /nix/store/bfgqhz5n5ad83h9bqxzvhz158fb03vyz-SDKs/MacOSX10.12.sdk -enable-objc-interop -emit-objc-header -parse-as-library -O -D HAVE_MACOS_10_11_FEATURES -module-name macOS_swift -emit-module-path generated/osdep/macOS_swift.swiftmodule -import-objc-header /private/tmp/nix-build-mpv-0.35.1.drv-0/source/osdep/macOS_swift_bridge.h -emit-objc-header-path generated/osdep/macOS_swift.h -o generated/osdep/macOS_swift.o ../osdep/macos/libmpv_helper.swift ../osdep/macos/log_helper.swift ../osdep/macos/mpv_helper.swift ../osdep/macos/swift_compat.swift ../osdep/macos/swift_extensions.swift ../video/out/mac/common.swift ../video/out/mac/title_bar.swift ../video/out/mac/view.swift ../video/out/mac/window.swift ../video/out/cocoa_cb_common.swift ../video/out/mac/gl_layer.swift ../osdep/macos/remote_command_center.swift -I. -I/private/tmp/nix-build-mpv-0.35.1.drv-0/source
../osdep/macos/libmpv_helper.swift:18:8: error: no such module 'Cocoa'
import Cocoa
       ^
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
```

After(https://logs.nix.ci/?key=nixos/nixpkgs.214944&attempt_id=76802592-760c-47e1-9141-4465b7c78f4c):
```
[21/452] Generating generated/osdep/swift_targets with a custom command
FAILED: generated/osdep/macOS_swift.swiftmodule generated/osdep/macOS_swift.h generated/osdep/macOS_swift.o 
/nix/store/5mldrl986vi6b8lk3gdw8r5mjxqx5i9j-swift-wrapper-5.7.3/bin/swift -frontend -c -sdk /nix/store/y421982rvq7fwfcygxhwala06fxq4sdm-SDKs/MacOSX11.0.sdk -enable-objc-interop -emit-objc-header -parse-as-library -O -D HAVE_MACOS_10_11_FEATURES -D HAVE_MACOS_10_14_FEATURES -module-name macOS_swift -emit-module-path generated/osdep/macOS_swift.swiftmodule -import-objc-header /private/tmp/nix-build-mpv-0.35.1.drv-0/source/osdep/macOS_swift_bridge.h -emit-objc-header-path generated/osdep/macOS_swift.h -o generated/osdep/macOS_swift.o ../osdep/macos/libmpv_helper.swift ../osdep/macos/log_helper.swift ../osdep/macos/mpv_helper.swift ../osdep/macos/swift_compat.swift ../osdep/macos/swift_extensions.swift ../video/out/mac/common.swift ../video/out/mac/title_bar.swift ../video/out/mac/view.swift ../video/out/mac/window.swift ../video/out/cocoa_cb_common.swift ../video/out/mac/gl_layer.swift ../osdep/macos/remote_command_center.swift -I. -I/private/tmp/nix-build-mpv-0.35.1.drv-0/source
<module-includes>:1:9: note: in file included from <module-includes>:1:
#import "Headers/OpenGL.h"
        ^
/nix/store/7i2mvhisb2wnjcvc9jvzlxn78mlsiazq-apple-framework-OpenGL-11.0.0/Library/Frameworks/OpenGL.framework/Headers/OpenGL.h:31:147: error: pasting formed '__MAC_10.0', an invalid preprocessing token
extern CGLError CGLChoosePixelFormat(const CGLPixelFormatAttribute *attribs, CGLPixelFormatObj OPENGL_NULLABLE * OPENGL_NONNULL pix, GLint *npix) OPENGL_DEPRECATED(10.0, 10.14);
                                                                                                                                                  ^
/nix/store/xgb5hzizdmrh83672gxn5llx74vqz0r3-apple-framework-OpenGL/Library/Frameworks/OpenGL.framework/Headers/OpenGLAvailability.h:11:74: note: expanded from macro 'OPENGL_DEPRECATED'
#define OPENGL_DEPRECATED(from, to) __OSX_AVAILABLE_BUT_DEPRECATED(__MAC_##from, __MAC_##to, __IPHONE_NA, __IPHONE_NA)
                                                                         ^
```

I can reproduce the same error using `nix build .#legacyPackages.x86_64-darwin.mpv` on a aarch64-darwin machine.

Atemu reviewed (approved):
Result of `nixpkgs-review pr 214944` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mpv</li>
  </ul>
</details>

Diff LGTM and it works just fine on aarch64-darwin.

Closure size optimisation can come later, that's not a hard blocker IMO.

Once the x86 issue is fixed, this is good to go.

ylh commented on pkgs/applications/video/mpv/default.nix:
```diff
@@ -66,7 +66,7 @@
 , sixelSupport       ? false,          libsixel
 , speexSupport       ? true,           speex
-, swiftSupport       ? false,          swift
+, swiftSupport       ? stdenv.isDarwin, swift
```
Might it be worth doing this for the time being and merging?

```suggestion
, swiftSupport       ? stdenv.isDarwin && stdenv.isAarch64, swift
```

azuwis replied:
Done in https://github.com/NixOS/nixpkgs/commit/8a28ae851dee5306de51352300fceb692bd9c389

Atemu commented on pkgs/applications/video/mpv/default.nix:
```diff
@@ -99,6 +99,11 @@ in stdenv.mkDerivation (self: {
 
+  preConfigure = lib.optionalString swiftSupport ''
+    # Ensure we reference 'lib' (not 'out') of Swift.
+    export SWIFT_LIB_DYNAMIC=${swift.swift.lib}/lib/swift/macosx
```
```suggestion
    export SWIFT_LIB_DYNAMIC=${lib.getLib swift.swift}/lib/swift/macosx
```

azuwis replied:
Squash into f0a6216eebb3a6693b111e2927e10f520f6c479a

azuwis committed:
mpv: Only enable swift support on aarch64-darwin

This commit should be reverted after [x86_64-darwin build failure][1] is
fixed

[1]: https://logs.nix.ci/?key=nixos/nixpkgs.214944&attempt_id=76802592-760c-47e1-9141-4465b7c78f4c

Atemu reviewed (approved):
Result of `nixpkgs-review pr 214944` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>hydrus</li>
    <li>jellyfin-mpv-shim</li>
    <li>mpc-qt</li>
    <li>plex-mpv-shim</li>
    <li>sublime-music</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>mnemosyne</li>
    <li>subtitleedit</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>ani-cli</li>
    <li>anki</li>
    <li>cplay-ng</li>
    <li>curseradio</li>
    <li>dmlive</li>
    <li>mpv-unwrapped</li>
    <li>python310Packages.mpv</li>
    <li>python311Packages.mpv</li>
    <li>somafm-cli</li>
    <li>wtwitch</li>
    <li>ytfzf</li>
  </ul>
</details>

wegank commented:
Result of `nixpkgs-review pr 214944` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>hydrus</li>
    <li>jellyfin-mpv-shim</li>
    <li>mpc-qt</li>
    <li>plex-mpv-shim</li>
    <li>sublime-music</li>
  </ul>
</details>
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>ani-cli</li>
    <li>anki</li>
    <li>cplay-ng</li>
    <li>curseradio</li>
    <li>dmlive</li>
    <li>mnemosyne</li>
    <li>mpv-unwrapped</li>
    <li>python310Packages.mpv</li>
    <li>python311Packages.mpv</li>
    <li>somafm-cli</li>
    <li>subtitleedit</li>
    <li>wtwitch</li>
    <li>ytfzf</li>
  </ul>
</details>
